### PR TITLE
Feat(serverless-contracts): add SQS contract

### PR DESCRIPTION
**Pull request by CorentinDoue**
# SQS Contract

SQS is not easy to use. Messages are serialized and batched. This contract aims to simplify the usage of SQS. It abstracts serialization and deserialization of the messages while providing type safety at runtime and in the IDE. It also abstracts technical batching process to write or read list of messages.

## Features

- `getSendMessage`: returns a function that **validates**, **serializes** and **send** a single message through SQS
- `getSendMessages`: returns a function that **validates**, **serializes**, **batches** and **send** a list of messages through SQS. Handle **retry** if the SQS is throttled because there are too many messages
- `getHandler`: returns a type safe lambda handler that **deserializes**, **validates**, **handle batched messages one by one** and **properly inform SQS of failed messages**. The callback (internal handler) receives one message and will be called in parallel for each message of the batch. If it throws, the lambda will not throw and only its message will be retuned to the SQS.
- `getLambdaHandler` (naming is bad, but is the same as AGW contract): returns a type safe lambda handler that **deserializes** and **validates** messages. The callback (internal handler) receives the whole batch of messages and is free to process them as it wants.

## TODO before merge

- [x] Publish alpha: https://www.npmjs.com/package/@swarmion/serverless-contracts/v/0.32.0-alpha.0
- [ ] e2e test the alpha on some projects
- [x] Add documentation: https://docs-preview-996--swarmion.netlify.app/docs/how-to-guides/use-serverless-contracts/sqs/

### CODE REVIEW

**Comment by CorentinDoue** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=swarmion&pullRequest=996) **Quality Gate failed**
> Failed conditions  [77.1% Coverage on New Code](https://sonarcloud.io/component_measures?id=swarmion&pullRequest=996&metric=new_coverage&view=list) (required ≥ 80%)
> 
> [See analysis details on SonarCloud](https://sonarcloud.io/dashboard?id=swarmion&pullRequest=996)

image
image

The coverage is destroyed by the types' folder/files that contains only types. Do you want / Can you exclude them from the sonar coverage?

**Comment by CorentinDoue** [comment]
@swarmion/core ⚠️ I think your dependabot is becoming crazy. I think all test (linter, types, ...) hits nx remote cache so all dependabots PR are green and it merges all of them.

For example, when I really execute the test-type command I can see that all vitest.config.ts are broken. Probably since https://github.com/swarmion/swarmion/commit/d71fd6174999ee702eb02988882488cc66396f37

image

**Comment by fargito** [comment]
@CorentinDoue my first guess would be that the lockfile isn't properly taken into account by Nx to compute its cache key. I'll try and look into this

**Comment by fargito** [comment]
@CorentinDoue I checked on `main` without any Nx cache and everything works as expected. I also tried to run tests on this branch with no cache and it breaks. Can you rebase your branch on main? I think it would solve the type issue

File: user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+```ts
+import { getHandler } from '@swarmion/serverless-contracts';
+import { ajv } from 'libs/ajv';
+
+export const main = getHandler(mySqsContract, {
+  ajv,
+  handleBatchedRecords: false,
+})(async ({ Records }) => {
+  Records.forEach(message => {
```
Comment (fargito) [inline user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:176]: Nit: the AWS API convention is a bit weird with its UpperCamelCase object keys, I would prefer to use `records`, WDYT?

File: user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    const { toto } = message.body; // parsed and typed with the correct keys
+
+    // write your business logic for one message
+  });
+  // Handle message failure as you want.
+  // Be aware that unhandled failure will make the whole batch of messages aivalable after the visbility timeout.
```
Comment (fargito) [inline user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:182]:
```suggestion
  // Be aware that unhandled failure will make the whole batch of messages available after the visibility timeout,
```

File: user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  validateMessage: false,
+});
+
+await sendMyMessage({ body: { what: 'ever' } });
+```
+
+:::
+
+:::caution
+This only works with the AWS sdk v3
```
Comment (fargito) [inline user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:264]:
```suggestion
This only works with the AWS SDK v3
```

File: swarmion.code-workspace
Comment (fargito) [inline swarmion.code-workspace:1]: I think these formatting changes will cause a prettier fail, can you run `pnpm format-fix` at the repo root?

File: packages/serverless-contracts/vitest.config.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ import { configDefaults, defineConfig } from 'vitest/config';
 export default defineConfig({
   plugins: [tsconfigPaths(), codspeedPlugin()],
   test: {
+    mockReset: true,
```
Comment (fargito) [inline packages/serverless-contracts/vitest.config.ts:8]: Curious: why is that necessary now?

File: packages/serverless-contracts/src/contracts/cloudFormation/__tests__/fullContractSchema.type.test.ts [hunk 5]
```diff
@@ -2,8 +2,9 @@ import { JSONSchema } from 'json-schema-to-ts';
 
 import { FullContractSchemaType } from '../types';
 
-type Check =
-  FullContractSchemaType<'hello'> extends JSONSchema ? 'pass' : 'fail';
+type Check = FullContractSchemaType<'hello'> extends JSONSchema
+  ? 'pass'
+  : 'fail';
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/cloudFormation/__tests__/fullContractSchema.type.test.ts:7]: I'm a little bit confused by this change, it should disappear with a `pnpm format-fix` at the repo root

File: packages/serverless-contracts/src/contracts/eventBridge/__tests__/fullContractSchema.type.test.ts [hunk 6]
```diff
@@ -9,10 +9,11 @@ const eventBridgeContract = new EventBridgeContract({
...
 
-type Check =
-  FullContractSchemaType<typeof eventBridgeContract> extends JSONSchema
-    ? 'pass'
-    : 'fail';
+type Check = FullContractSchemaType<
+  typeof eventBridgeContract
+> extends JSONSchema
+  ? 'pass'
+  : 'fail';
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/eventBridge/__tests__/fullContractSchema.type.test.ts:16]: I'm a little bit confused by this change, it should disappear with a `pnpm format-fix` at the repo root

File: packages/serverless-contracts/src/contracts/SQS/utils/serializeMessageAttributes.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    default:
+      throw new Error(
+        'Invalid messageAttributesSchema. Only string and number are currently supported',
+      );
+  }
+};
+export const serializeMessageAttributes = <
+  MessageAttributes extends Record<string, unknown>,
+  Contract extends SQSContract,
+>(
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/SQS/utils/serializeMessageAttributes.ts:33]: So this only supports a single depth of objects? If that is the case, maybe this should be mentioned in the docs

File: packages/serverless-contracts/src/contracts/SQS/utils/messagesValidation.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+export const validateMessage = <Contract extends SQSContract>({
+  contract,
+  message,
+  options,
+  index,
+}: {
+  contract: Contract;
+  message: SqsMessage<Contract>;
+  options: SendMessageBuilderOptions<Contract>;
+  index?: number;
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/SQS/utils/messagesValidation.ts:17]: Could this be required?

**Review by fargito — CHANGES_REQUESTED** [review]
This is awesome! I have a couple questions/suggestions, but this is impressive work

File: user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md (hunk 1 above)
Comment (CorentinDoue) [reply user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:176]: I agree, I changed it

**Comment by CorentinDoue** [comment]
Ok, 

> @swarmion/core ⚠️ I think your dependabot is becoming crazy. I think all test (linter, types, ...) hits nx remote cache so all dependabots PR are green and it merges all of them.
> 
> For example, when I really execute the test-type command I can see that all vitest.config.ts are broken. Probably since [d71fd61](https://github.com/swarmion/swarmion/commit/d71fd6174999ee702eb02988882488cc66396f37)
> 
> image
> 
> ➡️ it breaks my PR 😢

UPDATE: It was my PR, it was started a long time ago and when I rebased it on main with the new dependencies pnpm handled pretty bad the conflicts and ended downgrading some dependencies to the version they were when I started the PR. 

➡️ I fixed this with a rebase interactive, stopped at each commit that changed the `pnpm.lock`, for each of them, I reverted the broken lock and reinstalled the dependencies I changed in my commit from the package.json

File: swarmion.code-workspace
Comment (CorentinDoue) [reply swarmion.code-workspace:1]: I removed this commit after fixing the pnpm.lock (cf https://github.com/swarmion/swarmion/pull/996#issuecomment-2041456965)

File: packages/serverless-contracts/src/contracts/cloudFormation/__tests__/fullContractSchema.type.test.ts (hunk 5 above)
Comment (CorentinDoue) [reply packages/serverless-contracts/src/contracts/cloudFormation/__tests__/fullContractSchema.type.test.ts:7]: I removed this commit after fixing the pnpm.lock (cf https://github.com/swarmion/swarmion/pull/996#issuecomment-2041456965)

File: packages/serverless-contracts/src/contracts/eventBridge/__tests__/fullContractSchema.type.test.ts (hunk 6 above)
Comment (CorentinDoue) [reply packages/serverless-contracts/src/contracts/eventBridge/__tests__/fullContractSchema.type.test.ts:16]: I removed this commit after fixing the pnpm.lock (cf https://github.com/swarmion/swarmion/pull/996#issuecomment-2041456965)

File: packages/serverless-contracts/vitest.config.ts (hunk 4 above)
Comment (CorentinDoue) [reply packages/serverless-contracts/vitest.config.ts:8]: It's not necessary, but I prefer to have the mocks reset before each test. It avoids side effects of one test on another. Otherwise, lots of `beforeEach` are needed to explicitly reset the mocks when some side effects are discovered, but some can be forgotten.

It's especially useful when counting mock calls. You want all the calls of the previous test to have been forgotten

File: packages/serverless-contracts/src/contracts/SQS/utils/messagesValidation.ts (hunk 8 above)
Comment (CorentinDoue) [reply packages/serverless-contracts/src/contracts/SQS/utils/messagesValidation.ts:17]: No, the`validateMessage` is used in both `sendMessage` and `sendMessages`. In `sendMessage` there is only one message, so there is no `index`. 

`index` is used here only to print better error messages

**Comment by CorentinDoue** [comment]
@fargito I fixed the PR and acted on your comments

**Comment by CorentinDoue** [comment]
[RFC] I want to e2e test the contract. Actually this could be done with integration tests. I need your advice/help on some points

1. I want to create a new CDK service `services/contracts-tests` that creates some lambdas that use each contract. Each lambda has some integration test that ensure the contract works as expected. **WDYT?**
2. I can deploy and run the test on one of my account for the development purpose but it could be cool to deploy this service on the staging env and run the test after the deployment as it was done with the integration tests of the orchestrator service. I noticed the staging deployment is broken for 5 month and the integration tests are skipped. I think I probably shut down the account where they were deployed. What do you to do with this? Do you want to remove the staging or is there a plan to get back a dedicated account and fix the CD?

**Review by fargito — APPROVED** [review]
Great work! Thanks a lot.
As for the integration tests:
- I agree with the new service. Maybe we can simply rename the `orchestrator` service and use its infrastructure, since it used to be the source of integration tests
- I'm not really sure we need two separate accounts for prod and staging. We could simply run the integration tests on production after merging on main

Anyway, let's continue this discussion in #525 as on a broader scale, it could also serve as a demonstration of cloud-native integration testing good practices.

**Comment by fargito** [comment]
@allcontributors please add @CorentinDoue for docs

**Comment by fargito** [comment]
I need to fix #1143 before releasing the beta for 0.32.0, but I'll try and do it fast
